eval.go

```diff
@@ -450,26 +466,47 @@ func gatherMatches(mt matchTree, known map[matchTree]bool) []*candidateMatch {
 	}
 	cands = res
 
-	// Merge adjacent candidates. This guarantees that the matches
-	// are non-overlapping.
-	sort.Sort((sortByOffsetSlice)(cands))
-	res = cands[:0]
-	for i, c := range cands {
-		if i == 0 {
+	if merge {
+		// Merge adjacent candidates. This guarantees that the matches
+		// are non-overlapping.
+		sort.Sort((sortByOffsetSlice)(cands))
+		res = cands[:0]
+		for i, c := range cands {
+			if i == 0 {
+				res = append(res, c)
+				continue
+			}
+			last := res[len(res)-1]
+			lastEnd := last.byteOffset + last.byteMatchSz
+			end := c.byteOffset + c.byteMatchSz
+			if lastEnd >= c.byteOffset {
+				if end > lastEnd {
+					last.byteMatchSz = end - last.byteOffset
+				}
+				continue
+			}
+
 			res = append(res, c)
-			continue
 		}
-		last := res[len(res)-1]
-		lastEnd := last.byteOffset + last.byteMatchSz
-		end := c.byteOffset + c.byteMatchSz
-		if lastEnd >= c.byteOffset {
-			if end > lastEnd {
-				last.byteMatchSz = end - last.byteOffset
+	} else {
+		// Remove overlapping candidates. This guarantees that the matches
+		// are non-overlapping, but also preserves expected match counts.
+		// TODO do we actually need this guarantee?
+		sort.Sort((sortByOffsetSlice)(cands))
+		res = cands[:0]
+		for i, c := range cands {
+			if i == 0 {
+				res = append(res, c)
+				continue
+			}
+			last := res[len(res)-1]
+			lastEnd := last.byteOffset + last.byteMatchSz
+			if lastEnd > c.byteOffset {
+				continue
 			}
-			continue
-		}
 
-		res = append(res, c)
+			res = append(res, c)
+		}
 	}
```

Comment:
Small hack here: previously, we were merging ranges that overlapped or were adjacent. We don't want that because we want our match count to be accurate. Searching for `(a|b)` on the text `ab` should return two matches, not one.

This just discards any matches that are strictly overlapping, but allows adjacent matches to continue to exist. Note that this new behavior is behind the `merge` argument, which is only set when `opts.ChunkMatches == true`.